allennlp/data/semparse/worlds/world.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict
+import pyparsing
+
+from nltk.sem.logic import Expression, LambdaExpression, BasicType, Type
+
+from allennlp.data.semparse.type_declarations.type_declaration import DynamicTypeLogicParser
+
+
+class World:
+    """
+    Base class for defining a world in a new domain. This class defines a method to translate a logical form
+    as per a naming convention that works with NLTK's ``LogicParser``. The sub-classes can decide on the
+    convention by overriding the ``_map_name`` method that does token level mapping. This class also defines
+    methods for transforming logical form strings into parsed ``Expressions``, and ``Expressions`` into
+    action sequences.
+
+    Parameters
+    ----------
+    constant_type_prefixes : ``Dict[str, BasicType]`` (optional)
+        If you have an unbounded number of constants in your domain, you are required to add prefixes to their
+        names to denote their types. This is the mapping from prefixes to types.
+    global_type_signatures : ``Dict[str, Type]`` (optional)
+        A mapping from translated names to their types.
+    global_name_mapping : ``Dict[str, str]`` (optional)
+        A name mapping from the original names in the domain to the translated names.
+    """
+    def __init__(self,
+                 constant_type_prefixes: Dict[str, BasicType] = None,
+                 global_type_signatures: Dict[str, Type] = None,
+                 global_name_mapping: Dict[str, str] = None) -> None:
+        # NLTK has a naming convention for variable types. If the world has predicate or entity names beyond
+        # what's defined in the COMMON_NAME_MAPPING, they need to be added to this dict.
+        # We initialize this dict with common predicate names and update it as we process logical forms.
+        self.local_name_mapping: Dict[str, str] = {}
+        # Similarly, these are the type signatures not in the COMMON_TYPE_SIGNATURE.
+        self.local_type_signatures: Dict[str, Type] = {}
+        self.global_name_mapping = global_name_mapping or {}
+        self.global_type_signatures = global_type_signatures or {}
+        # We keep a reverse map as well to put the terminals back in action sequences.
+        self.reverse_name_mapping = {mapped_name: name for name, mapped_name in self.global_name_mapping.items()}
+        type_prefixes = constant_type_prefixes or {}
+        self._logic_parser = DynamicTypeLogicParser(constant_type_prefixes=type_prefixes,
+                                                    type_signatures=self.global_type_signatures)
+
+    def parse_logical_form(self, logical_form: str) -> Expression:
+        """
+        Takes a logical form as a string, maps its tokens using the mapping and returns a parsed expression.
+        """
+        def _process_nested_expression(nested_expression) -> str:
```

Comment:
This should not be a nested function.  In general, the only reason to have a nested function is if you're _returning_ the function itself.  Otherwise you're almost certainly better off defining the function without nesting, both for code readability and for efficiency.

Reply:
Done.